published/external/xdp/details/apiioctl.h

```diff
@@ -0,0 +1,386 @@
+#define IOCTL_XSK_NOTIFY_ASYNC \
+    CTL_CODE(FILE_DEVICE_NETWORK, 5, METHOD_NEITHER, FILE_WRITE_ACCESS)
+
+//
+// Input struct for IOCTL_XSK_BIND
+//
+typedef struct _XSK_BIND_IN {
+    UINT32 IfIndex;
+    UINT32 QueueId;
+    XSK_BIND_FLAGS Flags;
+} XSK_BIND_IN;
+
+//
+// Input struct for IOCTL_XSK_ACTIVATE
+//
+typedef struct _XSK_ACTIVATE_IN {
+    XSK_ACTIVATE_FLAGS Flags;
+} XSK_ACTIVATE_IN;
+
+//
+// Input struct for IOCTL_XSK_SET_SOCKOPT
+//
+typedef struct _XSK_SET_SOCKOPT_IN {
+    UINT32 Option;
+    UINT32 InputBufferLength;
+    const VOID *InputBuffer;
+} XSK_SET_SOCKOPT_IN;
+
+//
+// Input struct for IOCTL_XSK_NOTIFY
+//
+typedef struct _XSK_NOTIFY_IN {
+    XSK_NOTIFY_FLAGS Flags;
+    UINT32 WaitTimeoutMilliseconds;
+} XSK_NOTIFY_IN;
+
+#ifndef _KERNEL_MODE
+
+//
+// This struct is defined in public kernel headers, but not user mode headers.
+//
+typedef struct _FILE_FULL_EA_INFORMATION {
+    ULONG NextEntryOffset;
+    UCHAR Flags;
+    UCHAR EaNameLength;
+    USHORT EaValueLength;
+    CHAR EaName[1];
+} FILE_FULL_EA_INFORMATION;
+
+#endif
+
+#define XDP_OPEN_EA_LENGTH \
+    (sizeof(FILE_FULL_EA_INFORMATION) + sizeof(XDP_OPEN_PACKET_NAME) + sizeof(XDP_OPEN_PACKET))
+
+inline
+XDP_STATUS
+_XdpCloseHandle(
+    _In_ HANDLE Handle
+    )
+{
+#ifdef _KERNEL_MODE
+    return ZwClose(Handle);
+#else
+    return CloseHandle(Handle) ?
+        XDP_STATUS_SUCCESS : HRESULT_FROM_WIN32(RtlNtStatusToDosError(GetLastError()));
+#endif
+}
+
+inline
+XDP_STATUS
+_XdpConvertNtStatusToXdpStatus(
+    _In_ NTSTATUS Status
+    )
+{
+#ifdef _KERNEL_MODE
+    return Status;
+#else
+    if (Status != STATUS_SUCCESS) {
+        return HRESULT_FROM_WIN32(RtlNtStatusToDosError(Status));
+    }
+#endif
+}
+
+inline
+XDP_STATUS
+_XdpCreateEvent(
+    _Out_ HANDLE *EventHandle
+    )
+{
+#ifdef _KERNEL_MODE
+    return ZwCreateEvent(EventHandle, EVENT_ALL_ACCESS, NULL, NotificationEvent, FALSE);
+#else
+    *EventHandle = CreateEventW(NULL, FALSE, FALSE, NULL);
+    if (*EventHandle == NULL) {
+        return HRESULT_FROM_WIN32(RtlNtStatusToDosError(GetLastError()));
+    }
+#endif
+}
+
+inline
+XDP_STATUS
+_XdpWaitInfinite(
+    _In_ HANDLE *EventHandle
+    )
+{
+#ifdef _KERNEL_MODE
+    return ZwWaitForSingleObject(EventHandle, FALSE, NULL);
+#else
+    DWORD WaitResult = WaitForSingleObject(EventHandle, INFINITE);
+    if (WaitResult == WAIT_OBJECT_0) {
+        return XDP_STATUS_SUCCESS;
+    } else if (WaitResult == WAIT_FAILED) {
+        return HRESULT_FROM_WIN32(RtlNtStatusToDosError(GetLastError()));
+    } else {
+        return XDP_STATUS_INTERNAL_ERROR;
+    }
+#endif
+}
+
+inline
+VOID *
+_XdpInitializeEaVersion(
+    _In_ XDP_OBJECT_TYPE ObjectType,
+    _In_ UINT32 ApiVersion,
+    _Out_ VOID *EaBuffer,
+    _In_ ULONG EaLength
+    )
+{
+    FILE_FULL_EA_INFORMATION *EaHeader = EaBuffer;
+    XDP_OPEN_PACKET *OpenPacket;
+
+    if (EaLength < XDP_OPEN_EA_LENGTH) {
+        __fastfail(FAST_FAIL_INVALID_ARG);
```

Comment:
We should make all error-handling within this header configurable/overrideable by the includer.

This particular line should probably be converted to an `XDPAPI_ASSERT`, which the caller can predefine to whatever it wants.